kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md

```diff
@@ -238,71 +238,184 @@ A cons-list is used for the EVM wordstack.
-
-    syntax Bool ::= Int "in" WordStack [function]
- // ---------------------------------------------
-    rule _ in .WordStack => false
-    rule W in (W' : WS)  => (W ==K W') orElseBool (W in WS)
-```
-
--   `#replicateAux` pushes `N` copies of `A` onto a `WordStack`.
--   `#replicate` is a `WordStack` of length `N` with `A` the value of every element.
-
-```k
-    syntax WordStack ::= #replicate    ( Int, Int )            [symbol(#replicate), function, total]
-                       | #replicateAux ( Int, Int, WordStack ) [symbol(#replicateAux), function, total]
- // ---------------------------------------------------------------------------------------------------
-    rule #replicate   ( N,  A )     => #replicateAux(N, A, .WordStack)
-    rule #replicateAux( N,  A, WS ) => #replicateAux(N -Int 1, A, A : WS) requires         N >Int 0
-    rule #replicateAux( N, _A, WS ) => WS                                 requires notBool N >Int 0
+    syntax Int ::= #sizeWordStack ( WordStack ) [symbol(#sizeWordStack), function, total, smtlib(sizeWordStack)]
+ // ------------------------------------------------------------------------------------------------------------
+    rule [ws-size-00]: #sizeWordStack (.WordStack) => 0
+    rule [ws-size-01]: #sizeWordStack (_ : .WordStack) => 1
+    rule [ws-size-02]: #sizeWordStack (_ : _ : .WordStack) => 2
+    rule [ws-size-03]: #sizeWordStack (_ : _ : _ : .WordStack) => 3
+    rule [ws-size-04]: #sizeWordStack (_ : _ : _ : _ : .WordStack) => 4
+    rule [ws-size-05]: #sizeWordStack (_ : _ : _ : _ : _ : .WordStack) => 5
+    rule [ws-size-06]: #sizeWordStack (_ : _ : _ : _ : _ : _ : .WordStack) => 6
+    rule [ws-size-07]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : .WordStack) => 7
+    rule [ws-size-08]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 8
+    rule [ws-size-09]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 9
+    rule [ws-size-10]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 10
+    rule [ws-size-11]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 11
+    rule [ws-size-12]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 12
+    rule [ws-size-13]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 13
+    rule [ws-size-14]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 14
+    rule [ws-size-15]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 15
+    rule [ws-size-16]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 16
+    rule [ws-size-N]:  #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : WS) => 16 +Int #sizeWordStack(WS)
+    rule [ws-size-O]:  #sizeWordStack (_ : WS) => 1 +Int #sizeWordStack(WS) [priority(75)]
```

Comment:
I think the reason for how it was structured before was to keep the computation tail-recursive, which the LLVM backend requires for efficient memory/time usage. Can we check the LLVM backend performance on teh conformance test-suite here?

Reply:
I don't like this, I've reverted back to just the simple tail recursion.

@ehildenb Do you remember why there was an auxiliary function here with two parameters instead of a direct computation?

Reply:
https://github.com/runtimeverification/evm-semantics/blob/3d7c9c2b9419a663af0f7e3e8dfd4da50ee16892/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md?plain=1#L398-L401

This isn't tail recursive. It's doing an add operation after the call. That's why there was an auxiliary function with the extra parameter.

Reply:
Ah yes yes, sorry.

Reply:
I will bring it back. It's not the thing that's breaking the proofs anyways and works well.

If we wanted to speed this calculation up, we can merge something like [this](https://github.com/runtimeverification/evm-semantics/pull/2628).